samples/bluetooth/gatt/nrf_dfu.c

```diff
@@ -0,0 +1,388 @@
+			rsp[2] = RSP_OP_FAILED;
+		} else {
+			/* Go back to started state */
+			state = DFU_STATE_STARTED;
+		}
+
+		/* Check device_type if any */
+		if ((device_type != 0) && (device_type != init->device_type)) {
+			rsp[2] = RSP_OP_FAILED;
+		}
+
+		/* Check device_revision if any */
+		if ((device_revision != 0) &&
+		    (device_revision != init->device_revision)) {
+			rsp[2] = RSP_OP_FAILED;
+		}
+
+		/* do not care about softDevice, CRC1-16-CCIT is last field */
+		crc16 = sys_le16_to_cpu(init->softdevice_id[init->softdevice_id_count]);
+
+		bt_gatt_notify(conn,  &nrf_dfu_attrs[1], &rsp, sizeof(rsp));
+
+		if (rsp[2] != RSP_SUCCESS) {
+			reset_state(conn);
+		}
+
+		break;
+	}
+	case DFU_STATE_RECV:
+	{
+		int err = 0;
+
+		if (cb && cb->receive) {
+			err = cb->receive(img_offset, buf, len);
+		}
+
+		exp_size -= len;
+		img_offset += len;
+		pkt_count++;
+
+		if (pkt_notif && !(pkt_count % pkt_notif)) {
+			u8_t rsp[] = { DFU_OP_PKT_NOTIF,
+				       0x00, 0x00, 0x00, 0x00 };
+			u32_t *bytes = (u32_t *)&rsp[2];
+
+			*bytes = sys_cpu_to_le32(img_offset);
+
+			/* Report Received Image Size */
+			bt_gatt_notify(conn,  &nrf_dfu_attrs[1], &rsp,
+				       sizeof(rsp));
+		}
+
+		if (exp_size == 0 || err) {
+			u8_t rsp[] = { DFU_OP_RSP_CODE, DFU_OP_RECV_IMG,
+				       err ? RSP_OP_FAILED : RSP_SUCCESS };
+
+			state = DFU_STATE_RECV_DONE;
+
+			bt_gatt_notify(conn,  &nrf_dfu_attrs[1], &rsp,
+				       sizeof(rsp));
+
+			if (rsp[2] != RSP_SUCCESS) {
+				reset_state(conn);
+			}
+		}
+
+		break;
+	}
+	default:
+		return -EINVAL;
+	}
+
+	return len;
+}
+
+static ssize_t read_version(struct bt_conn *conn,
+			    const struct bt_gatt_attr *attr,
+			    void *buf, u16_t len, u16_t offset)
+{
+	u16_t version;
+
+	if (cb && cb->reset_to_boot && state != DFU_STATE_BOOTLOADER) {
+		version = 0x0001; /* 0.1 (app mode) */
+	} else {
+		version = 0x0004; /* 0.4 (bootloader mode) */
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &version,
+				 sizeof(version));
+}
+
+static struct bt_gatt_ccc_cfg control_ccc_cfg[BT_GATT_CCC_MAX] = {};
+
+static struct bt_gatt_attr nrf_dfu_attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(&nrf_dfu_uuid),
+	BT_GATT_CHARACTERISTIC(&nrf_dfu_control.uuid,
+			       BT_GATT_CHRC_WRITE | BT_GATT_CHRC_NOTIFY,
+			       BT_GATT_PERM_WRITE,
+			       NULL, write_control, NULL),
+	BT_GATT_CCC(control_ccc_cfg, NULL),
+	BT_GATT_CHARACTERISTIC(&nrf_dfu_packet.uuid,
+			       BT_GATT_CHRC_WRITE_WITHOUT_RESP,
+			       BT_GATT_PERM_WRITE,
+			       NULL, write_packet, NULL),
+	BT_GATT_CHARACTERISTIC(&nrf_dfu_version.uuid,
+			       BT_GATT_CHRC_READ,
+			       BT_GATT_PERM_READ,
+			       read_version, NULL, NULL),
+};
+
+static struct bt_gatt_service nrf_dfu_svc = BT_GATT_SERVICE(nrf_dfu_attrs);
+
+void nrf_dfu_init(struct nrf_dfu_cb *callbacks, u16_t type, u16_t revision)
+{
+	cb = callbacks;
+	device_type = type;
+	device_revision = revision;
+
+	bt_gatt_service_register(&nrf_dfu_svc);
```

Comment:
If you don't intend to unregister then it might be better registering as a static service with BT_GATT_SERVICE_DEFINE.

Reply:
ok thx will do